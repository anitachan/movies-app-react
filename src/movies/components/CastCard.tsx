import { Box, Button, Card, CardMedia, Typography } from '@mui/material';
import { Cast } from '../models/cast.model';
import noUser from '../../assets/no-user.png';
import { useNavigate } from 'react-router-dom';

interface Props {
  actor: Cast;
}
export const CastCard = ({ actor }: Props) => {
  const url = process.env.REACT_APP_IMAGES_URL;
  const profile_path = actor.profile_path ? `${url}${actor.profile_path}` : noUser;
  const navigate = useNavigate();

  const handleClick = () => {
    navigate(`/actor/${actor.id!}`);
  };

  return (
    <Button onClick={handleClick}>
      <Box sx={{ maxWidth: 250, textAlign: 'center' }}>
        <Card sx={{ maxWidth: 250 }}>
          <CardMedia
            component="img"
            src={profile_path}
            alt={actor.name}
            sx={{
              width: '100%',
              height: 350,
              objectFit: 'cover',
              borderRadius: 2,
            }}
          />
        </Card>
        <Typography variant="body2" fontWeight="bold" noWrap>
          {actor.name}
        </Typography>
        <Typography variant="caption" color="text.secondary" noWrap>
          {actor.character}
        </Typography>
      </Box>
    </Button>
  );
};
