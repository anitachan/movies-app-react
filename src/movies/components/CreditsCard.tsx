import { Box, Button, Typography } from '@mui/material';
import noUser from '../../assets/no-user.png';
import { ActorCast } from '../models/actorCast.model';
import { useNavigate } from 'react-router-dom';

interface Props {
	credit: ActorCast;
}

export const CreditsCard = ({ credit }: Props) => {
	const url = process.env.REACT_APP_IMAGES_URL;
	const posterSrc = credit.poster_path ? `${url}${credit.poster_path}` : noUser;
	const navigate = useNavigate();

	const handleClick = () => {
		navigate(`/details/${credit.id!}`);
	};

	return (
		<Button onClick={handleClick}>
			<Box sx={{ width: 150, textAlign: 'center' }}>
				<Box
					component="img"
					src={posterSrc}
					alt={credit.title}
					loading="lazy"
					onError={(e) => {
						e.currentTarget.onerror = null;
						e.currentTarget.src = noUser;
					}}
					sx={{
						width: '100%',
						height: 225, // ratio 2:3
						objectFit: 'cover',
						borderRadius: 1,
						display: 'block',
					}}
				/>
				<Typography variant="subtitle2" noWrap>
					{credit.title}
				</Typography>
				<Typography variant="caption" color="text.secondary" noWrap>
					{credit.release_date ? new Date(credit.release_date).getFullYear() : 'â€”'} as {credit.character}
				</Typography>
			</Box>
		</Button>
	);
};
