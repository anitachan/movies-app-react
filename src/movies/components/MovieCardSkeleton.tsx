import { Box, Skeleton } from '@mui/material';

export const MovieCardSkeleton = () => (
	<Box sx={{ position: 'relative' }}>
		<Skeleton variant="rectangular" width="100%" height={200} sx={{ borderRadius: 4 }} />
		<Box
			sx={{
				bgcolor: 'rgba(0, 0, 0, 0.7)',
				color: 'white',
				p: 2,
				mt: '-4px',
				display: 'flex',
				flexDirection: 'row',
				justifyContent: 'space-between',
				alignItems: 'center',
				borderBottomLeftRadius: 4,
				borderBottomRightRadius: 4,
			}}
		>
			<Box sx={{ flex: 1 }}>
				<Skeleton variant="text" width="80%" />
				<Skeleton variant="text" width="100%" />
				<Skeleton variant="text" width="40%" />
			</Box>
			<Skeleton variant="rectangular" width={80} height={30} sx={{ borderRadius: 1, ml: 2 }} />
		</Box>
	</Box>
);
