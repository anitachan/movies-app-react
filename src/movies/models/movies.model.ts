export interface NowPlaying {
  page: number,
  results: Movie[],
  total_pages: number,
  total_results: number,
}

export interface Movie {
  popularity?: number;
  vote_count?: number;
  video?: boolean;
  poster_path: string;
  id: number;
  adult?: boolean;
  backdrop_path: string;
  original_language?: OriginalLanguage;
  original_title?: string;
  genre_ids?: number[];
  title: string;
  vote_average: number;
  overview: string;
  release_date?: string;
  poster_url?: string;
}

export enum OriginalLanguage {
  En = 'en',
  Es = 'es',
  Ko = 'ko',
}
