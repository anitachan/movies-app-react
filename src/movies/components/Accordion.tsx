import { ExpandMore } from '@mui/icons-material';
import LaunchIcon from '@mui/icons-material/Launch';
import {
  Accordion as MUIAccordion,
  AccordionDetails,
  AccordionSummary,
  Avatar,
  Typography,
  Chip,
  Stack,
  Rating,
} from '@mui/material';
import { MovieDetail } from '../models/movieDetail';

interface Props {
  movie: MovieDetail;
}
export const Accordion = ({ movie }: Props) => {
  return (
    <>
      <MUIAccordion defaultExpanded>
        <AccordionSummary
          expandIcon={<ExpandMore />}
          aria-controls="panel1-content"
          id="panel1-header"
        >
          <Typography component="span">Synopsis</Typography>
        </AccordionSummary>
        <AccordionDetails>
          <Typography>{movie?.overview}</Typography>
        </AccordionDetails>
      </MUIAccordion>
      <MUIAccordion defaultExpanded>
        <AccordionSummary
          expandIcon={<ExpandMore />}
          aria-controls="panel2-content"
          id="panel2-header"
        >
          <Typography component="span">Budget</Typography>
        </AccordionSummary>
        <AccordionDetails>
          <Typography>$ {movie?.budget}</Typography>
        </AccordionDetails>
      </MUIAccordion>
      <MUIAccordion defaultExpanded>
        <AccordionSummary
          expandIcon={<ExpandMore />}
          aria-controls="panel2-content"
          id="panel2-header"
        >
          <Typography component="span">Genres</Typography>
        </AccordionSummary>
        <AccordionDetails>
          {movie?.genres.map((genre) => (
            <Chip
              key={genre.id}
              label={genre.name}
              avatar={<Avatar> {genre.name.substring(0, 1)} </Avatar>}
            />
          ))}
        </AccordionDetails>
      </MUIAccordion>
      <MUIAccordion defaultExpanded>
        <AccordionSummary
          expandIcon={<ExpandMore />}
          aria-controls="panel2-content"
          id="panel2-header"
        >
          <Typography component="span">Production Companies</Typography>
        </AccordionSummary>
        <AccordionDetails>
          {movie?.production_companies.map((companied) => (
            <Chip
              key={companied.id}
              label={companied.name}
              avatar={<Avatar src={companied.logo_path!} />}
            />
          ))}
        </AccordionDetails>
      </MUIAccordion>
      <MUIAccordion defaultExpanded>
        <AccordionSummary
          expandIcon={<ExpandMore />}
          aria-controls="panel2-content"
          id="panel2-header"
        >
          <Typography component="span">HomePage</Typography>
        </AccordionSummary>
        <AccordionDetails>
          <a
            href={movie?.homepage}
            target="_blank"
            rel="noopener noreferrer"
            style={{
              display: 'flex',
              alignItems: 'center',
              gap: 4,
              textDecoration: 'none',
              color: 'inherit',
            }}
          >
            <LaunchIcon fontSize="small" />
            <Typography>{movie?.homepage}</Typography>
          </a>
        </AccordionDetails>
      </MUIAccordion>
      <MUIAccordion defaultExpanded>
        <AccordionSummary
          expandIcon={<ExpandMore />}
          aria-controls="panel2-content"
          id="panel2-header"
        >
          <Typography component="span">Rating</Typography>
        </AccordionSummary>
        {movie?.vote_average && (
          <AccordionDetails>
            <Stack direction="row" spacing={1} alignItems="center">
              <Rating value={movie?.vote_average} precision={0.1} max={10} readOnly />
              <Typography variant="body2">{movie?.vote_average.toPrecision(2)}</Typography>
            </Stack>
          </AccordionDetails>
        )}
      </MUIAccordion>
    </>
  );
};
