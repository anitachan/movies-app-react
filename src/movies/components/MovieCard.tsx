import { Box, Button, Rating, Stack, Typography } from '@mui/material';
import { Movie } from '../models/movies.model';
import { useNavigate } from 'react-router-dom';
import { MovieDetail } from '../models/movieDetail';

interface Props {
  movie: Movie | MovieDetail;
}

export const MovieCard = ({ movie }: Props) => {
  const url = process.env.REACT_APP_IMAGES_URL;
  const { backdrop_path, title, overview, vote_average } = movie;
  const navigate = useNavigate();

  const handleClick = () => {
    navigate(`/details/${movie.id}`);
  };

  return (
    <Box sx={{ position: 'relative' }}>
      <img
        src={`${url}${backdrop_path}?w=400&fit=crop&auto=format`}
        alt={title}
        loading="lazy"
        style={{
          width: '100%',
          height: 'auto',
          objectFit: 'cover',
          display: 'block',
          borderRadius: 4,
        }}
      />
      <Box
        sx={{
          bgcolor: 'rgba(0, 0, 0, 0.7)',
          color: 'white',
          p: 2,
          mt: '-4px',
          display: 'flex',
          flexDirection: 'row',
          justifyContent: 'space-between',
          alignItems: 'center',
          borderBottomLeftRadius: 4,
          borderBottomRightRadius: 4,
        }}
      >
        <Box>
          <Typography
            sx={{
              display: '-webkit-box',
              WebkitLineClamp: 1,
              WebkitBoxOrient: 'vertical',
              overflow: 'hidden',
              textOverflow: 'ellipsis',
            }}
            variant="subtitle1"
            fontWeight="bold"
          >
            {title}
          </Typography>
          <Typography
            sx={{
              display: '-webkit-box',
              WebkitLineClamp: 2,
              WebkitBoxOrient: 'vertical',
              overflow: 'hidden',
              textOverflow: 'ellipsis',
            }}
            variant="body2"
          >
            {overview}
          </Typography>
          <Stack direction="row" spacing={1} alignItems="center">
            <Rating value={vote_average} precision={0.1} max={10} readOnly />
            <Typography variant="body2">{vote_average.toPrecision(2)}</Typography>
          </Stack>
        </Box>
        <Button
          sx={{ marginLeft: '10px', whiteSpace: 'nowrap' }}
          variant="contained"
          size="small"
          color="primary"
          onClick={handleClick}
        >
          See More
        </Button>
      </Box>
    </Box>
  );
};
