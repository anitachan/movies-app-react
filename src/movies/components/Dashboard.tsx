import { Box } from '@mui/material';
import { useInfiniteScroll } from '../hooks/useInfiniteScroll';
import { Movie } from '../models/movies';
import { MovieCard } from './MovieCard';
import { MovieCardSkeleton } from './MovieCardSkeleton';

interface Props {
	movies: Movie[];
	loadMore: () => void;
}

export const Dashboard = ({ movies, loadMore }: Props) => {
	const loaderRef = useInfiniteScroll(loadMore);

	return (
		<>
			<Box
				sx={{
					display: 'grid',
					minHeight: '120vh',
					gridTemplateColumns: {
						xs: 'repeat(1, 1fr)',
						sm: 'repeat(2, 1fr)',
						md: 'repeat(3, 1fr)',
						lg: 'repeat(4, 1fr)',
					},
					gap: 2,
				}}
			>
				{movies.length > 0
					? movies.map((movie) => <MovieCard key={movie.id} movie={movie} />)
					: [...Array(9)].map((_, i) => <MovieCardSkeleton key={i} />)}
			</Box>
			{movies.length > 0 && <div ref={loaderRef} data-testid="loader" style={{ height: '60px' }} />}
		</>
	);
};
