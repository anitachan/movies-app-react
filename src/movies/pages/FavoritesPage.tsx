import { Dashboard } from '../components/Dashboard';
import { useFavorites } from '../hooks/useFavorites';

export const FavoritesPage = () => {
  const { favorites } = useFavorites();

  return <Dashboard movies={favorites!} />;
};
