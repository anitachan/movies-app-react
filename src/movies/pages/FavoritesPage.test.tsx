import { render } from '@testing-library/react';
import { FavoritesPage } from './FavoritesPage';
import { Dashboard } from '../components/Dashboard';
import { fixtureOf } from 'ts-mock-autofixture-kit';
import { MovieDetail } from '../models/movieDetail';

const mockUseNavigate = jest.fn();
jest.mock(
  'react-router-dom',
  () => ({
    useNavigate: () => mockUseNavigate,
  }),
  { virtual: true },
);
const mockMovies = fixtureOf<MovieDetail[]>({ seed: 42 }).create();

jest.mock('../hooks/useFavorites', () => ({
  useFavorites: () => ({ favorites: mockMovies }),
}));

jest.mock('../components/Dashboard', () => ({
  __esModule: true,
  Dashboard: jest.fn(() => <div data-testid="mock-dashboard">Dashboard Mock</div>),
}));

describe('FavoritesPage', () => {
  test('should render Favorite Page', () => {
    const { container } = render(<FavoritesPage />);

    expect(container).toMatchSnapshot();
    expect(Dashboard).toHaveBeenCalledWith(
      expect.objectContaining({ movies: mockMovies }),
      undefined,
    );
  });
});
