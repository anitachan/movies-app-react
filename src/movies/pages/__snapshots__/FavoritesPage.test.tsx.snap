// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesPage should render Favorite Page 1`] = `<div />`;
